File: Visual Studio LIghtswitch 2012/Chapter13/Chapter13/ApressExtensionVB/ApressExtensionVB/ApressExtensionVB.Client/Presentation/Shells/ScreenWrapper.vb
'You can use and redistribute the code from this book (and sample application) for non-commercial and 
'most commercial purposes as long as you acknowledge the source and authorship. 
'You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
'The acknowledgment should include author, title, publisher, and ISBN. 
'An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2012 by Tim Leung, published by Apress, ISBN 9781430250715".

Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Linq
Imports System.Windows
Imports Microsoft.LightSwitch
Imports Microsoft.LightSwitch.Client
Imports Microsoft.LightSwitch.Details
Imports Microsoft.LightSwitch.Details.Client
Imports Microsoft.LightSwitch.Utilities

Namespace Presentation.Shells

    Public Class ScreenWrapper
        Implements IScreenObject
        Implements INotifyPropertyChanged

        Private screenObject As IScreenObject
        Private dirty As Boolean
        Private dataServicePropertyChangedListeners As List(
           Of IWeakEventListener)

        Public Event PropertyChanged(sender As Object,
           e As PropertyChangedEventArgs) Implements INotifyPropertyChanged.PropertyChanged

        ' 1.  REGISTER FOR CHANGE NOTIFICATIONS
        Friend Sub New(screenObject As IScreenObject)
            Me.screenObject = screenObject
            Me.dataServicePropertyChangedListeners =
               New List(Of IWeakEventListener)

            ' Register for property changed events on the details object.
            AddHandler CType(screenObject.Details, 
              INotifyPropertyChanged).PropertyChanged,
              AddressOf Me.OnDetailsPropertyChanged

            ' Register for changed events on each of the data services.
            Dim dataServices As IEnumerable(Of IDataService) =
                screenObject.Details.DataWorkspace.Details.Properties.All().OfType(
                   Of IDataWorkspaceDataServiceProperty)().Select(
                      Function(p) p.Value)

            For Each dataService As IDataService In dataServices
                Me.dataServicePropertyChangedListeners.Add(
                  CType(dataService.Details, INotifyPropertyChanged).CreateWeakPropertyChangedListener(
                    Me, AddressOf Me.OnDataServicePropertyChanged))
            Next
        End Sub

        Private Sub OnDetailsPropertyChanged(sender As Object,
           e As PropertyChangedEventArgs)
            If String.Equals(e.PropertyName,
              "ValidationResults", StringComparison.OrdinalIgnoreCase) Then
                RaiseEvent PropertyChanged(
                   Me, New PropertyChangedEventArgs("ValidationResults"))
            End If
        End Sub

        Private Sub OnDataServicePropertyChanged(sender As Object,
          e As PropertyChangedEventArgs)
            Dim dataService As IDataService =
           CType(sender, IDataServiceDetails).DataService
            Me.IsDirty = dataService.Details.HasChanges
        End Sub

        ' 2.  EXPOSE AN ISDIRTY PROPERTY
        Public Property IsDirty As Boolean
            Get
                Return Me.dirty
            End Get
            Set(value As Boolean)
                Me.dirty = value
                RaiseEvent PropertyChanged(
                      Me, New PropertyChangedEventArgs("IsDirty"))
            End Set
        End Property

        ' 3.  EXPOSE A VALIDATION RESULTS PROPERTY
        Public ReadOnly Property ValidationResults As ValidationResults
            Get
                Return Me.screenObject.Details.ValidationResults
            End Get
        End Property

        ' 4.  EXPOSE UNDERLYING SCREEN PROPERTIES
        Public ReadOnly Property CanSave As Boolean Implements IScreenObject.CanSave
            Get
                Return Me.screenObject.CanSave
            End Get
        End Property

        Public Sub Close(promptUserToSave As Boolean) Implements IScreenObject.Close
            Me.screenObject.Close(promptUserToSave)
        End Sub

        Friend ReadOnly Property RealScreenObject As IScreenObject
            Get
                Return Me.screenObject
            End Get
        End Property

        Public Property Description As String Implements IScreenObject.Description
            Get
                Return Me.screenObject.Description
            End Get
            Set(value As String)
                Me.screenObject.Description = value
            End Set
        End Property

        Public ReadOnly Property Details As IScreenDetails Implements IScreenObject.Details
            Get
                Return Me.screenObject.Details
            End Get
        End Property

        Public Property DisplayName As String Implements IScreenObject.DisplayName
            Get
                Return Me.screenObject.DisplayName
            End Get
            Set(value As String)
                Me.screenObject.DisplayName = value
            End Set
        End Property

        Public ReadOnly Property Name As String Implements IScreenObject.Name
            Get
                Return Me.screenObject.Name
            End Get
        End Property

        Public Sub Refresh() Implements IScreenObject.Refresh
            Me.screenObject.Refresh()
        End Sub

        Public Sub Save() Implements IScreenObject.Save
            Me.screenObject.Save()
        End Sub

        Public ReadOnly Property Details1 As IBusinessDetails Implements IBusinessObject.Details
            Get
                Return CType(Me.screenObject, IBusinessObject).Details
            End Get
        End Property

        Public ReadOnly Property Details2 As IDetails Implements IObjectWithDetails.Details
            Get
                Return CType(Me.screenObject, IObjectWithDetails).Details
            End Get
        End Property

        Public ReadOnly Property Details3 As IStructuralDetails Implements IStructuralObject.Details
            Get
                Return CType(Me.screenObject, IStructuralObject).Details
            End Get
        End Property
    End Class

End Namespace

File: Visual Studio LIghtswitch 2012/Chapter13/Chapter13/ApressExtensionCS/ApressExtensionCS/ApressExtensionCS.Client/Presentation/Shells/ScreenWrapper.cs
//You can use and redistribute the code from this book (and sample application) for non-commercial and 
//most commercial purposes as long as you acknowledge the source and authorship. 
//You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
//The acknowledgment should include author, title, publisher, and ISBN. 
//An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2012 by Tim Leung, published by Apress, ISBN 9781430250715".

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Windows;

namespace ApressExtensionCS.Presentation.Shells
{
    using Microsoft.LightSwitch;
    using Microsoft.LightSwitch.Client;
    using Microsoft.LightSwitch.Details;
    using Microsoft.LightSwitch.Details.Client;
    using Microsoft.LightSwitch.Utilities;

    public class ScreenWrapper : IScreenObject, INotifyPropertyChanged
    {
        private IScreenObject screenObject;
        private bool dirty;
        private List<IWeakEventListener> dataServicePropertyChangedListeners;

        public event PropertyChangedEventHandler PropertyChanged;

        // 1.  REGISTER FOR CHANGE NOTIFICATIONS
        internal ScreenWrapper(IScreenObject screenObject)
        {
            this.screenObject = screenObject;
            this.dataServicePropertyChangedListeners =
               new List<IWeakEventListener>();

            // Register for property changed events on the details object.
            ((INotifyPropertyChanged)screenObject.Details).PropertyChanged +=
                this.OnDetailsPropertyChanged;

            // Register for changed events on each of the data services.
            IEnumerable<IDataService> dataServices =
              screenObject.Details.DataWorkspace.Details.Properties.All().OfType<
               IDataWorkspaceDataServiceProperty>().Select(p => p.Value);

            foreach (IDataService dataService in dataServices)
                this.dataServicePropertyChangedListeners.Add(
                  ((INotifyPropertyChanged)dataService.Details).CreateWeakPropertyChangedListener(
                     this, this.OnDataServicePropertyChanged));
        }

        private void OnDetailsPropertyChanged(
           object sender, PropertyChangedEventArgs e)
        {
            if (String.Equals(
               e.PropertyName, "ValidationResults", StringComparison.OrdinalIgnoreCase))
            {
                if (null != this.PropertyChanged)
                    PropertyChanged(
                       this, new PropertyChangedEventArgs("ValidationResults"));
            }
        }

        private void OnDataServicePropertyChanged(
           object sender, PropertyChangedEventArgs e)
        {
            IDataService dataService = ((IDataServiceDetails)sender).DataService;
            this.IsDirty = dataService.Details.HasChanges;
        }

        // 2.  EXPOSE AN ISDIRTY PROPERTY
        public bool IsDirty
        {
            get { return this.dirty; }
            set
            {
                this.dirty = value;
                if (null != this.PropertyChanged)
                    PropertyChanged(
                      this, new PropertyChangedEventArgs("IsDirty"));
            }
        }

        // 3.  EXPOSE A VALIDATION RESULTS PROPERTY
        public ValidationResults ValidationResults
        {
            get { return this.screenObject.Details.ValidationResults; }
        }

        // 4.  EXPOSE UNDERLYING SCREEN PROPERTIES
        public IScreenDetails Details
        {
            get { return this.screenObject.Details; }
        }
        internal IScreenObject RealScreenObject
        {
            get { return this.screenObject; }
        }

        public string Name
        {
            get { return this.screenObject.Name; }
        }

        public string DisplayName
        {
            get { return this.screenObject.DisplayName; }
            set { this.screenObject.DisplayName = value; }
        }

        public string Description
        {
            get { return this.screenObject.Description; }
            set { this.screenObject.Description = value; }
        }

        public bool CanSave
        {
            get { return this.screenObject.CanSave; }
        }

        public void Save()
        {
            this.screenObject.Save();
        }

        public void Refresh()
        {
            this.screenObject.Refresh();
        }

        public void Close(bool promptUserToSave)
        {
            this.screenObject.Close(promptUserToSave);
        }

        IBusinessDetails IBusinessObject.Details
        {
            get { return ((IBusinessObject)this.screenObject).Details; }
        }

        IStructuralDetails IStructuralObject.Details
        {
            get { return ((IStructuralObject)this.screenObject).Details; }
        }

        IDetails IObjectWithDetails.Details
        {
            get { return ((IObjectWithDetails)this.screenObject).Details; }
        }
    }
}
